spaceteam_proxy_handler.cpp

```diff
@@ -0,0 +1,364 @@
+		if (tokens[0] == "host") {
+			if (tokens.size() >= 2) {
+				this->host = tokens[1];
+			}
+		} else if (tokens[0] == "proxy_port") {
+			if (tokens.size() >= 2) {
+				this->port = tokens[1];
+			}
+		} else {
+			return bad_request;
+		}
+	}
+
+	return OK;
+}
+
+
+RequestHandler::Status SpaceteamProxyHandler::HandleRequest(const Request& req, Response* res){
+	std::cout << host << std::endl;
+	if (res == nullptr) {
+		return bad_request;
+	}
+
+	while (true) {
+		// connect to host
+		boost::asio::io_service io_service;
+		boost::asio::ip::tcp::resolver resolver(io_service);
+		boost::asio::ip::tcp::resolver::query query(host, port);
+		boost::asio::ip::tcp::resolver::iterator endpoint_iterator;
+		try {
+			endpoint_iterator = resolver.resolve(query);
+		} catch (std::exception& e) {
+			std::cerr << "Exception: " << e.what() << std::endl;
+			NotFoundHandler not_found;
+			not_found.HandleRequest(req, res);
+			return OK;
+		}
+
+		boost::asio::ip::tcp::socket sock(io_service);
+		boost::asio::connect(sock, endpoint_iterator);
+
+		// get relative URI
+		std::string request_uri = req.uri();
+		std::string relative_uri;
+		if (request_uri != uri 
+				&& uri != "/"
+				&& request_uri.size() > uri.size() 
+				&& request_uri.substr(0, uri.size()) == uri) {
+			relative_uri = request_uri.substr(uri.size());
+		} else {
+			relative_uri = request_uri;
+		}
+
+		// send request to host
+		std::string request = "GET " + relative_uri + " HTTP/1.1\r\n"
+												+ "Host: " + host + ":" + port + "\r\n"
+												+ "Connection: close\r\n\r\n";
+		boost::asio::write(sock, 
+											 boost::asio::buffer(request, request.length()));
+
+		// get response from host
+		boost::asio::streambuf buffer;
+		boost::system::error_code error;
+		while (boost::asio::read(sock, buffer, error)) {
+			if (error) break;
+		}
+		std::string raw_response((std::istreambuf_iterator<char>(&buffer)), std::istreambuf_iterator<char>());
+		ParseResponse(raw_response);
+
+		if (response_code == 301 || response_code == 302) { // redirect. try new host
+			for (auto h : headers) {
+				if (h.first == "Location") {
+					host = h.second;
+				}
+			}
+		} else {
+			std::cerr << "sent response for " << relative_uri << std::endl;
+			break;
+		}
+	}
+
+	res->SetStatus(static_cast<Response::ResponseCode>(response_code));
+	for (auto h : headers) {
+		res->AddHeader(h.first, h.second);
+	}
+	res->SetBody(body);
+
+	return OK;
+}
+
+
+RequestHandler::Status SpaceteamProxyHandler::ParseResponse(const std::string& raw_response){
+	// clear response member variables
+	http_version = "";
+	response_code = 0;
+	headers.clear();
+	body.clear();
+	state = http_version_h;
+
+	// parse raw response, character by character
+	for (char c : raw_response) {
+		result_type result = consume(c);
+		if (result == result_type::bad) {
+			return bad_request;
+		}
+	}
+	return OK;
+}
+
+
+SpaceteamProxyHandler::result_type SpaceteamProxyHandler::consume(char input){

```

Comment:
It's not duplicated! Our's is different from yours, though a decent amount is the same.